regression-tests.recursor-dnssec/test_Flags.py

```diff
@@ -0,0 +1,679 @@
+import socket
+import unittest
+
+import dns
+from recursortests import RecursorTest
+
+class TestFlags(RecursorTest):
+    _confdir = 'Flags'
+    _config_template = """dnssec=%s"""
+    _config_params = ['_dnssec_setting']
+    _dnssec_setting = None
+    _recursors = {}
+
+    _dnssec_setting_ports = {'off': 5300, 'process': 5301, 'validate': 5302}
+
+    @classmethod
+    def setUp(cls):
+        for setting in cls._dnssec_setting_ports:
+            confdir = os.path.join('configs', cls._confdir, setting)
+            cls.wipeRecursorCache(confdir)
+
+    @classmethod
+    def setUpClass(cls):
+        cls.setUpSockets()
+        confdir = os.path.join('configs', cls._confdir)
+        cls.createConfigDir(confdir)
+
+        cls.generateAllAuthConfig(confdir)
+        cls.startAllAuth(confdir)
+
+        for dnssec_setting, port in cls._dnssec_setting_ports.items():
+            cls._dnssec_setting = dnssec_setting
+            recConfdir = os.path.join(confdir, dnssec_setting)
+            cls.createConfigDir(recConfdir)
+            cls.generateRecursorConfig(recConfdir)
+            cls.startRecursor(recConfdir, port)
+            cls._recursors[dnssec_setting] = cls._recursor
+
+    @classmethod
+    def setUpSockets(cls):
+        cls._sock = {}
+        for dnssec_setting, port in cls._dnssec_setting_ports.items():
+            print("Setting up UDP socket..")
+            cls._sock[dnssec_setting] = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            cls._sock[dnssec_setting].settimeout(2.0)
+            cls._sock[dnssec_setting].connect(("127.0.0.1", port))
+
+    @classmethod
+    def sendUDPQuery(cls, query, dnssec_setting, timeout=2.0):
+        if timeout:
+            cls._sock[dnssec_setting].settimeout(timeout)
+
+        try:
+            cls._sock[dnssec_setting].send(query.to_wire())
+            data = cls._sock[dnssec_setting].recv(4096)
+        except socket.timeout:
+            data = None
+        finally:
+            if timeout:
+                cls._sock[dnssec_setting].settimeout(None)
+
+        msg = None
+        if data:
+            msg = dns.message.from_wire(data)
+        return msg
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.tearDownAuth()
+        for _, recursor in cls._recursors.items():
+            cls._recursor = recursor
+            cls.tearDownRecursor()
+
+    def createQuery(self, name, rdtype, flags, ednsflags):
+        """Helper function that creates the query with the specified flags.
+        The flags need to be strings (no checking is performed atm)"""
+        msg = dns.message.make_query(name, rdtype)
+        msg.flags = dns.flags.from_text(flags)
+        msg.flags += dns.flags.from_text('RD')
+        msg.use_edns(edns=0, ednsflags=dns.flags.edns_from_text(ednsflags))
+        return msg
+
+    def getQueryForSecure(self, flags='', ednsflags=''):
+        return self.createQuery('ns1.example.', 'A', flags, ednsflags)
+
+    ##
+    #   -AD -CD -DO
```

Comment:
This could be awesome, would require a multidimensional list (dnssec setting, domain status, flags (possibly split by flag on/off))